xdem/coreg.py

```diff
@@ -1063,3 +1081,620 @@ def coregister(reference_raster: Union[str, gu.georaster.Raster], to_be_aligned_
+
+    def _apply_func(self, dem: np.ndarray, transform: rio.transform.Affine) -> np.ndarray:
+        """Apply the coregistration steps sequentially to a DEM."""
+        dem_mod = dem.copy()
+
+        for coreg in self.pipeline:
+            dem_mod = coreg.apply(dem_mod, transform)
+
+        return dem_mod
+
+    def _apply_pts_func(self, coords: np.ndarray) -> np.ndarray:
+        """Apply the coregistration steps sequentially to a set of points."""
+        coords_mod = coords.copy()
+
+        for coreg in self.pipeline:
+            coords_mod = coreg._apply_pts_func(coords_mod)
+
+        return coords_mod
+
+    def _to_matrix_func(self) -> np.ndarray:
+        """Try to join the coregistration steps to a single transformation matrix."""
+        if not _HAS_P3D:
+            raise ValueError("Optional dependency needed. Install 'pytransform3d'")
+
+        transform_mgr = TransformManager()
+
+        with warnings.catch_warnings():
+            # Deprecation warning from pytransform3d. Let's hope that is fixed in the near future.
+            warnings.filterwarnings("ignore", message="`np.float` is a deprecated alias for the builtin `float`")
+            for i, coreg in enumerate(self.pipeline):
+                new_matrix = coreg.to_matrix()
+
+                transform_mgr.add_transform(i, i + 1, new_matrix)
+
+            return transform_mgr.get_transform(0, len(self.pipeline))
+
+    def __iter__(self):
+        """Iterate over the pipeline steps."""
+        for coreg in self.pipeline:
+            yield coreg
+
+    def __add__(self, other: Union[list[Coreg], Coreg, CoregPipeline]) -> CoregPipeline:
+        """Append Coreg(s) or a CoregPipeline to the pipeline."""
+        if not isinstance(other, Coreg):
+            other = list(other)
+        else:
+            other = [other]
+
+        pipelines = self.pipeline + other
+
+        return CoregPipeline(pipelines)
+
+
+class NuthKaab(Coreg):
+    """
+    Nuth and Kääb (2011) DEM coregistration.
+
+    Implemented after the paper:
+    https://doi.org/10.5194/tc-5-271-2011
+    """
+
+    def __init__(self, max_iterations: int = 50, error_threshold: float = 0.05):  # pylint: disable=super-init-not-called
+        """
+        Instantiate a new Nuth and Kääb (2011) coregistration object.
+
+        :param max_iterations: The maximum allowed iterations before stopping.
+        :param error_threshold: The residual change threshold after which to stop the iterations.
+        """
+        self.max_iterations = max_iterations
+        self.error_threshold = error_threshold
+
+        self._meta: dict[str, Any] = {}
+
+    def _fit_func(self, ref_dem: np.ndarray, tba_dem: np.ndarray, transform: Optional[rio.transform.Affine],
+                  weights: Optional[np.ndarray]):
+        """Estimate the x/y/z offset between two DEMs."""
+        bounds, resolution = _transform_to_bounds_and_res(ref_dem, transform)
+        verbose = False  # TODO: Make this an argument somewhere.
```

Comment:
Should there be an attribute in Coreg called verbose, that could be passed to each function?

Reply:
Added `verbose`  argument and associated callbacks in 1f3e735